source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -303,6 +358,20 @@ void SslSocket::shutdownSsl() {
   }
 }
 
+void SslSocket::asyncCb() {
+  ENVOY_CONN_LOG(debug, "SSL async done!", callbacks_->connection());
+
+  // We lose the return value here, so might consider propagating it with an event
+  // in case we run into "Close" result from the handshake handler.
+  if (state_ != SocketState::HandshakeComplete) {
+    PostIoAction action = doHandshake();
+    if (action == PostIoAction::Close) {
```

Comment:
There should probably be a test to see if handling the error cases (both in the OpenSSL engine and here) works right.

Reply:
Added a test